File: buildSrc/src/main/java/com/github/retrooper/strategy/JsonObjectCompressionStrategy.java
/*
 * This file is part of packetevents - https://github.com/retrooper/packetevents
 * Copyright (C) 2024 retrooper and contributors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.github.retrooper.strategy;

import com.github.difflib.DiffUtils;
import com.github.difflib.algorithm.DiffAlgorithmI;
import com.github.difflib.algorithm.myers.MeyersDiff;
import com.github.difflib.patch.*;
import com.github.retrooper.CompressionUtil;
import com.github.retrooper.EntryVersion;
import com.github.steveice10.opennbt.tag.builtin.*;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonPrimitive;

import java.util.*;

public class JsonObjectCompressionStrategy extends JsonCompressionStrategy {

    public static final JsonObjectCompressionStrategy INSTANCE = new JsonObjectCompressionStrategy();

    private JsonObjectCompressionStrategy() {
    }

    @Override
    void serialize(final CompoundTag tag, final JsonObject json) {
        final TreeMap<EntryVersion, JsonElement> entries = separateVersions(json);
        // we will remove the first entry, which is the oldest version
        final Map.Entry<EntryVersion, JsonElement> oldestVersion = entries.pollFirstEntry();
        tag.put("start", new StringTag(oldestVersion.getKey().toString()));

        final CompoundTag nbtEntries = new CompoundTag();
        Map<String, JsonPrimitive> ent = CompressionUtil.getObjectEntriesAsMap(oldestVersion.getValue().getAsJsonObject());

        final CompoundTag oldestVersionTag = new CompoundTag();
        for (final Map.Entry<String, JsonPrimitive> e : ent.entrySet()) {
            oldestVersionTag.put(e.getKey(), getAsNbtTag(e.getValue()));
        }
        nbtEntries.put(oldestVersion.getKey().toString(), oldestVersionTag);

        for (Map.Entry<EntryVersion, JsonElement> e : entries.entrySet()) {
            final Map<String, JsonPrimitive> map = CompressionUtil.getObjectEntriesAsMap(e.getValue().getAsJsonObject());
            final DiffAlgorithmI<Map.Entry<String, JsonPrimitive>> algo = MeyersDiff.factory().create(Map.Entry::equals);
            final Patch<Map.Entry<String, JsonPrimitive>> diff = DiffUtils.diff(new ArrayList<>(ent.entrySet()), new ArrayList<>(map.entrySet()), algo, null, false);

            final CompoundTag versionTag = new CompoundTag();
            final CompoundTag additions = new CompoundTag();
            final CompoundTag removals = new CompoundTag();
            for (AbstractDelta<Map.Entry<String, JsonPrimitive>> delta : diff.getDeltas()) {
                if (delta instanceof InsertDelta) {
                    for (Map.Entry<String, JsonPrimitive> line : delta.getTarget().getLines()) {
                        additions.put(line.getKey(), getAsNbtTag(line.getValue()));
                    }
                } else if (delta instanceof DeleteDelta) {
                    for (Map.Entry<String, JsonPrimitive> line : delta.getSource().getLines()) {
                        removals.put(line.getKey(), getAsNbtTag(line.getValue()));
                    }
                } else if (delta instanceof ChangeDelta) {
                    for (Map.Entry<String, JsonPrimitive> line : delta.getSource().getLines()) {
                        removals.put(line.getKey(), getAsNbtTag(line.getValue()));
                    }
                    for (Map.Entry<String, JsonPrimitive> line : delta.getTarget().getLines()) {
                        additions.put(line.getKey(), getAsNbtTag(line.getValue()));
                    }
                }
            }

            versionTag.put("additions", additions);
            versionTag.put("removals", removals);
            nbtEntries.put(e.getKey().toString(), versionTag);

            ent = map;
        }

        tag.put("entries", nbtEntries);
    }

    private Tag getAsNbtTag(final JsonPrimitive primitive) {
        // Will only be string or int as of the mappings
        if (primitive.isNumber()) {
            return new IntTag(primitive.getAsInt());
        } else if (primitive.isString()) {
            return new StringTag(primitive.getAsString());
        } else {
            throw new IllegalArgumentException("Unknown primitive type: " + primitive);
        }
    }

}

File: buildSrc/src/main/kotlin/com/github/retrooper/compression/strategy/JsonObjectCompressionStrategy.kt
/*
 * This file is part of packetevents - https://github.com/retrooper/packetevents
 * Copyright (C) 2024 retrooper and contributors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.github.retrooper.compression.strategy

import com.github.difflib.DiffUtils
import com.github.difflib.algorithm.myers.MeyersDiff
import com.github.difflib.patch.ChangeDelta
import com.github.difflib.patch.DeleteDelta
import com.github.difflib.patch.InsertDelta
import com.github.retrooper.compression.asPrimitiveMap
import com.github.steveice10.opennbt.tag.builtin.*
import com.google.gson.JsonObject
import com.google.gson.JsonPrimitive

object JsonObjectCompressionStrategy : JsonCompressionStrategy() {

    override fun serialize(tag: CompoundTag, json: JsonObject) {
        val entries = separateVersions(json)

        tag.put("length", ByteTag(entries.size.toByte()))

        // we will remove the first entry, which is the oldest version
        val oldestVersion = entries.pollFirstEntry()

        val nbtEntries = CompoundTag()
        var ent = oldestVersion.value.asJsonObject.asPrimitiveMap

        val oldestVersionTag = CompoundTag()
        for ((key, value) in ent) {
            oldestVersionTag.put(key, value.asNbtTag)
        }
        nbtEntries.put(oldestVersion.key.toString(), oldestVersionTag)

        for ((key, value) in entries) {
            val map = value.asJsonObject.asPrimitiveMap
            val algo = MeyersDiff.factory().create<Map.Entry<String, JsonPrimitive>> { a, b -> a == b }
            val diff = DiffUtils.diff(ArrayList(ent.entries), ArrayList(map.entries), algo, null, false)

            val versionTag = CompoundTag()
            val additions = CompoundTag()
            val removals = CompoundTag()
            for (delta in diff.deltas) {
                when (delta) {
                    is InsertDelta<*> -> {
                        for ((k, v) in delta.getTarget().lines) {
                            additions.put(k, v.asNbtTag)
                        }
                    }

                    is DeleteDelta<*> -> {
                        for ((k, v) in delta.getSource().lines) {
                            removals.put(k, v.asNbtTag)
                        }
                    }

                    is ChangeDelta<*> -> {
                        for ((k, v) in delta.getSource().lines) {
                            removals.put(k, v.asNbtTag)
                        }
                        for ((k, v) in delta.getTarget().lines) {
                            additions.put(k, v.asNbtTag)
                        }
                    }
                }
            }

            versionTag.put("removals", removals)
            versionTag.put("additions", additions)
            nbtEntries.put(key.toString(), versionTag)

            ent = map
        }

        tag.put("entries", nbtEntries)
    }

    private val JsonPrimitive.asNbtTag: Tag get() = if (isNumber) {
        IntTag(asInt)
    } else if (isString) {
        StringTag(asString)
    } else {
        throw IllegalArgumentException("Unknown primitive type: $this")
    }

}
